libgadget/exchange.c

```diff
@@ -63,54 +78,77 @@ int domain_exchange(int (*layoutfunc)(int p)) {
         MPI_Type_commit(&MPI_TYPE_PLAN_ENTRY);
     }
 
-    /*! toGo[0][task*NTask + partner] gives the number of particles in task 'task'
-     *  that have to go to task 'partner'
-     *  toGo[1] is SPH, toGo[2] is BH and toGo[3] is stars
-     */
     /* flag the particles that need to be exported */
     ExchangePlan plan;
-    plan.toGo = (ExchangePlanEntry *) mymalloc2("toGo", sizeof(plan.toGo[0]) * NTask);
-    plan.toGoOffset = (ExchangePlanEntry *) mymalloc2("toGo", sizeof(plan.toGo[0]) * NTask);
-    plan.toGet = (ExchangePlanEntry *) mymalloc2("toGet", sizeof(plan.toGo[0]) * NTask);
-    plan.toGetOffset = (ExchangePlanEntry *) mymalloc2("toGet", sizeof(plan.toGo[0]) * NTask);
+    /*Build a list of particles that will be exchanged*/
+    plan.first = 0;
+    plan.last = 0;
+    plan.ExchangeList = mymalloc2("exchangelist", sizeof(int) * PartManager->NumPart * omp_get_max_threads());
+    memset(plan.ExchangeList, 0, sizeof(int) * PartManager->NumPart * omp_get_max_threads());
+    size_t *nexthr = ta_malloc("nexthr", size_t, omp_get_max_threads());
```

Comment:
feels like this can be made into a function in the same spirit as gadget_compact_thread_arrays.

Reply:
void gadget_setup_thread_arrays(int * dest, int * srcs[], size_t sizes[], size_t total_size, int narrays)
{
    memset(dest, 0, sizeof(int) * total_size * narrays);
    srcs[0] = dest;
    for(i=0; i < narrays; i++) {
        srcs[i] = dest + i * total_size;
        sizes[i] = 0;
    }
}

Like that?

Reply:
Yes. -- Don't we also use this elsewhere?